File: examples/tailcat-e2b/js/src/e2bSandbox.ts
import { CommandExitError, Sandbox } from "e2b";
import { TAILCAT_URL, TEMPLATE_ALIAS } from "../template/template";

const INSTALL_TAILCAT_CMD =
  `curl -fsSL -o /tmp/tailcat.tgz ${TAILCAT_URL} && mkdir -p /tmp/tc && tar xzf /tmp/tailcat.tgz -C /tmp/tc` +
  " && sudo install -m 0755 $(find /tmp/tc -type f -name tailcat) /usr/local/bin/tailcat && rm -rf /tmp/tc /tmp/tailcat.tgz";

export interface CommandResult {
  exitCode: number;
  output: string;
}

export type CommandRunner = (command: string) => Promise<CommandResult>;

/** Create a sandbox from the Tailcat template, with a runtime-install fallback. */
export async function createSandbox(timeoutMs = 600_000): Promise<Sandbox> {
  const startedAt = Date.now();
  let sandbox: Sandbox;

  try {
    sandbox = await Sandbox.create(TEMPLATE_ALIAS, { timeoutMs });
  } catch (error) {
    if (!String(error).includes("404")) {
      throw error;
    }

    sandbox = await Sandbox.create({ timeoutMs });
    await sandbox.commands.run(INSTALL_TAILCAT_CMD, { timeoutMs: 120_000 });
    console.log(
      `[${sandbox.sandboxId}] template '${TEMPLATE_ALIAS}' not found; installed Tailcat at runtime`,
    );
  }

  const elapsedSeconds = (Date.now() - startedAt) / 1000;
  console.log(
    `[${sandbox.sandboxId}] sandbox ready in ${elapsedSeconds.toFixed(1)}s`,
  );
  return sandbox;
}

/** Run a sandbox command and preserve non-zero exit details for assertions. */
export async function runSandboxCommand(
  sandbox: Sandbox,
  command: string,
  timeoutMs = 60_000,
): Promise<CommandResult> {
  try {
    const result = await sandbox.commands.run(command, { timeoutMs });
    return { exitCode: result.exitCode, output: result.stdout + result.stderr };
  } catch (error) {
    if (error instanceof CommandExitError) {
      return { exitCode: error.exitCode, output: error.stdout + error.stderr };
    }
    throw error;
  }
}

export function sandboxCommandRunner(
  sandbox: Sandbox,
  timeoutMs = 120_000,
): CommandRunner {
  return (command) => runSandboxCommand(sandbox, command, timeoutMs);
}

export function assertCommandSucceeded(result: CommandResult): void {
  if (result.exitCode !== 0) {
    throw new Error(
      `command failed (exit ${result.exitCode}):\n${result.output}`,
    );
  }
}

File: examples/tailcat-e2b/python/tailcat_e2b/e2b_sandbox.py
"""Small E2B adapters shared by the Tailcat demos."""

from dataclasses import dataclass
import time
from typing import Callable

from e2b import CommandExitException, Sandbox

from .template import TAILCAT_URL, TEMPLATE_ALIAS

INSTALL_TAILCAT_CMD = (
    f"curl -fsSL -o /tmp/tailcat.tgz {TAILCAT_URL} && mkdir -p /tmp/tc && tar xzf /tmp/tailcat.tgz -C /tmp/tc"
    " && sudo install -m 0755 $(find /tmp/tc -type f -name tailcat) /usr/local/bin/tailcat"
    " && rm -rf /tmp/tc /tmp/tailcat.tgz"
)


@dataclass
class CommandResult:
    exit_code: int
    output: str


CommandRunner = Callable[[str], CommandResult]


def create_sandbox(timeout_seconds: int = 600) -> Sandbox:
    """Create a sandbox from the Tailcat template, with a runtime-install fallback."""
    started_at = time.time()
    try:
        sandbox = Sandbox.create(TEMPLATE_ALIAS, timeout=timeout_seconds)
    except Exception as error:
        if "404" not in str(error):
            raise

        sandbox = Sandbox.create(timeout=timeout_seconds)
        sandbox.commands.run(INSTALL_TAILCAT_CMD, timeout=120)
        print(f"[{sandbox.sandbox_id}] template '{TEMPLATE_ALIAS}' not found; installed Tailcat at runtime")

    print(f"[{sandbox.sandbox_id}] sandbox ready in {time.time() - started_at:.1f}s")
    return sandbox


def run_sandbox_command(sandbox: Sandbox, command: str, timeout_seconds: int = 60) -> CommandResult:
    """Run a sandbox command and preserve non-zero exit details for assertions."""
    try:
        result = sandbox.commands.run(command, timeout=timeout_seconds)
        return CommandResult(result.exit_code, result.stdout + result.stderr)
    except CommandExitException as error:
        return CommandResult(error.exit_code, error.stdout + error.stderr)


def sandbox_command_runner(sandbox: Sandbox, timeout_seconds: int = 120) -> CommandRunner:
    return lambda command: run_sandbox_command(sandbox, command, timeout_seconds)


def assert_command_succeeded(result: CommandResult) -> None:
    if result.exit_code != 0:
        raise RuntimeError(f"command failed (exit {result.exit_code}):\n{result.output}")
